diff --git a/nominatim/tokenizer/legacy_tokenizer.py b/nominatim/tokenizer/legacy_tokenizer.py
index 1b0b2980..1776006f 100644
--- a/nominatim/tokenizer/legacy_tokenizer.py
+++ b/nominatim/tokenizer/legacy_tokenizer.py
@@ -64,8 +64,9 @@ def _install_module(config_module_path: str, src_dir: Path, module_dir: Path) ->
         module_dir.mkdir()
 
     destfile = module_dir / 'nominatim.so'
-    shutil.copy(str(src_dir / 'nominatim.so'), str(destfile))
-    destfile.chmod(0o755)
+    if not destfile.is_file():
+        shutil.copy(str(src_dir / 'nominatim.so'), str(destfile))
+        destfile.chmod(0o755)
 
     LOG.info('Database module installed at %s', str(destfile))
 
