diff --git a/CMakeLists.txt b/CMakeLists.txt
index 536b21bc..3627d5e7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -138,6 +138,10 @@ if (BUILD_TESTS)
             set_tests_properties(bdd_${test}
                 PROPERTIES ENVIRONMENT "NOMINATIM_DIR=${PROJECT_BINARY_DIR}")
         endforeach()
+        # Do not run "bdd_{api,db,osm2pgsql}" in parallel with any other test.
+        set_tests_properties(bdd_api PROPERTIES RUN_SERIAL TRUE)
+        set_tests_properties(bdd_db PROPERTIES RUN_SERIAL TRUE TIMEOUT 2400)
+        set_tests_properties(bdd_osm2pgsql PROPERTIES RUN_SERIAL TRUE)
     else()
         message(WARNING "behave not found. BDD tests disabled." )
     endif()
diff --git a/nominatim/clicmd/setup.py b/nominatim/clicmd/setup.py
index 8464e151..f210b6df 100644
--- a/nominatim/clicmd/setup.py
+++ b/nominatim/clicmd/setup.py
@@ -67,7 +67,7 @@ class SetupAll:
                            help='Do not perform analyse operations during index (expert only)')
 
 
-    def run(self, args: NominatimArgs) -> int: # pylint: disable=too-many-statements
+    def run(self, args: NominatimArgs) -> int: # pylint: disable=too-many-branches, too-many-statements
         from ..data import country_info
         from ..tools import database_import, refresh, postcodes, freeze
         from ..indexer.indexer import Indexer
diff --git a/nominatim/config.py b/nominatim/config.py
index 3344a425..b02d6b3c 100644
--- a/nominatim/config.py
+++ b/nominatim/config.py
@@ -131,7 +131,7 @@ class Configuration:
     def get_str_list(self, name: str) -> Optional[List[str]]:
         """ Return the given configuration parameter as a list of strings.
             The values are assumed to be given as a comma-sparated list and
-            will be stripped before returning them. 
+            will be stripped before returning them.
 
             Parameters:
               name: Name of the configuration parameter with the NOMINATIM_
diff --git a/nominatim/server/falcon/server.py b/nominatim/server/falcon/server.py
index 5ec418a6..8382c90a 100644
--- a/nominatim/server/falcon/server.py
+++ b/nominatim/server/falcon/server.py
@@ -20,7 +20,7 @@ from nominatim.config import Configuration
 class HTTPNominatimError(Exception):
     """ A special exception class for errors raised during processing.
     """
-    def __init__(self, msg: str, status: int, content_type: str) -> None:
+    def __init__(self, msg: str, status: int, content_type: str) -> None: #pylint: disable=super-init-not-called
         self.msg = msg
         self.status = status
         self.content_type = content_type
diff --git a/test/bdd/api/reverse/geometry.feature b/test/bdd/api/reverse/geometry.feature
index 2c14dd5f..0f59d224 100644
--- a/test/bdd/api/reverse/geometry.feature
+++ b/test/bdd/api/reverse/geometry.feature
@@ -27,7 +27,7 @@ Feature: Geometries for reverse geocoding
           | 1            | 0.9               |
         Then results contain
           | geotext |
-          | POLYGON((9.5225302 47.138066,9.5225348 47.1379282,9.5227608 47.1379757,9.5227337 47.1380692,9.5225302 47.138066)) |
+          | POLYGON((9.5227337 47.1380692,9.5225348 47.1379282,9.5227608 47.1379757,9.5227337 47.1380692)) |
 
 
     Scenario: For polygons return the centroid as center point
