diff --git a/test/details.js b/test/details.js
index e543496..553929b 100644
--- a/test/details.js
+++ b/test/details.js
@@ -83,7 +83,7 @@ describe('Details Page', function () {
         assert.deepStrictEqual(display_headers, ['Name Keywords', 'Address Keywords']);
 
         let page_content = await page.$eval('body', el => el.textContent);
-        assert.ok(page_content.includes('qwaansshe')); // one of the name keywords
+        assert.ok(page_content.includes('vaduzo')); // one of the name keywords
       });
     }
 
