diff --git a/nominatim/clicmd/setup.py b/nominatim/clicmd/setup.py
index 344167bb..56256178 100644
--- a/nominatim/clicmd/setup.py
+++ b/nominatim/clicmd/setup.py
@@ -124,9 +124,12 @@ class SetupAll:
         tokenizer = self._get_tokenizer(args.continue_at, args.config)
 
         if args.continue_at is None or args.continue_at == 'load-data':
-            LOG.warning('Calculate postcodes')
-            postcodes.update_postcodes(args.config.get_libpq_dsn(),
-                                       args.project_dir, tokenizer)
+            if args.config.get_bool('NO_CALCULATE_POSTCODES'):
+                LOG.warning('Ignoring postcodes')
+            else:
+                LOG.warning('Calculate postcodes')
+                postcodes.update_postcodes(args.config.get_libpq_dsn(),
+                                           args.project_dir, tokenizer)
 
         if args.continue_at is None or args.continue_at in ('load-data', 'indexing'):
             LOG.warning('Indexing places')
diff --git a/settings/env.defaults b/settings/env.defaults
index 8378d7b0..e6c45177 100644
--- a/settings/env.defaults
+++ b/settings/env.defaults
@@ -42,6 +42,9 @@ NOMINATIM_LIMIT_REINDEXING=yes
 # The default is to not propagate these changes.
 NOMINATIM_UPDATE_FORWARD_DEPENDENCIES=no
 
+# If true, then postcodes won't be calculated on initial import.
+NOMINATIM_NO_CALCULATE_POSTCODES=no
+
 # Restrict search languages.
 # Normally Nominatim will include all language variants of name:XX
 # in the search index. Set this to a comma separated list of language
